src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/blob.py

```diff
@@ -0,0 +1,144 @@
+# ---------------------------------------------------------------------------------------------
+
+from __future__ import print_function
+from datetime import datetime
+from azure.cli.command_modules.storage._command_type import cli_storage_data_plane_command
+from azure.cli.command_modules.storage._factory import blob_data_service_factory
+from azure.storage.blob import BlockBlobService
+from azure.storage.blob.models import Include
+
+
+def storage_blob_copy(client, source_account, source_container, destination_container,
+                      source_sas=None,
+                      pattern=None, recursive=False, snapshots=False,
+                      exclude_old=False, exclude_new=False, blob_type="block"):
+    """
+    Copy blobs between containers and storage accounts. This is a server-side copy operation
+    therefore the command is asynchronous.
+
+    :param str source_account:
+        The account name of the source storage account
+
+    :param str source_container:
+        The source blob container
+
+    :param str source_sas:
+        The shared access signature used to access the source container. It is not required if
+        either a connection string is given or the source container doesn't require a sas.
+
+    :param str destination_container:
+        The destination blob container
+
+    :param str pattern:
+        If option --recursive is specified, then this command interprets the specified file pattern
+        as a blob prefix. If option --recursive is not specified, then this command matches the file
+        pattern against exact blob names.
+
+    :param bool recursive:
+        Copy all the files to the given container and maintain the folder structure.
+
+    :param bool snapshots:
+        Copy both the blobs and their snapshots.
+
+    :param bool exclude_old:
+        Excludes an older source resource. The resource will not be copied if the last modified time
+        of the source is the same or older than destination.
+
+    :param bool exclude_new:
+        Excludes a newer source resource. The resource will not be copied if the last modified time
+        of the source is the same or newer than destination.
+
+    :param str blob_type:
+        Specifies the type of the desination blob. the value can be block, page or append.
+
+    :return: A BlobCopyTicket instance summarize the operations
+    """
+
+    # TODO:
+    # 1. Page the result list (using num_result and marker)
+    # 2. Support connection string for source
+    # 3. stop using 'baseblobservice.exists' function. it doesn't provide performance gain
+    #    since it invoked the get_blob_properties any way.
+
+    # Question:
+    # 1. Performance of creating a source blob service
+
+    if pattern is None and not recursive:
+        # TODO: use a validator?
+        raise ValueError('Neither --recursive or --pattern is set.')
+
+    if exclude_old and exclude_new:
+        raise ValueError('Both --exclude-new and --exclude-old are set.')
```

Comment:
Why would this be an error? I would think this would be the equivalent of saying "Don't overwrite anything in dest that matches source; copy only things which don't match."